### PR TITLE
✨clusterctl use global logger

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:

As per discussion on https://github.com/kubernetes-sigs/cluster-api/pull/2151 (thanks @chuckha) this PR implements a global logger based on the controller runtime one (see the log.go file).

The PR validates the global logger approach by removing all the klog calls and implementing a   reviewed/finalized clusterctl init output.

After testing the two approaches discussed for logging, I'm +1 for this solution in clusterctl, but to complete the work we still need a custom logger...

**Which issue(s) this PR fixes**:
rif #1729

/area clusterctl
/assign @ncdc
/assign @vincepri
/assign @chuckha

### CODE REVIEW

File: cmd/clusterctl/pkg/client/init.go [hunk 1]
```diff
@@ -68,40 +71,53 @@ func (c *clusterctlClient) Init(options InitOptions) ([]Components, bool, error)
...
+	}
+
+	// IF this is the firstRun, then log the usage instructions.
+	if firstRun && options.LogUsageInstructions {
+		log.Info("")
+		log.Info("Your cluster API management cluster has been initialized successfully!")
+		log.Info("")
+		log.Info("You can now create your first workload cluster by running the following:")
+		log.Info("")
+		log.Info("  clusterctl config cluster [name] --kubernetes-version [version] | kubectl apply -f -")
```
Comment (detiber) [inline cmd/clusterctl/pkg/client/init.go]: This message seems out of place for the library, since it specifically calls out the binary name and arguments here.

File: cmd/clusterctl/pkg/log/log.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+// SetLogger sets a concrete logging implementation for all deferred Loggers.
+func SetLogger(l logr.Logger) {
+	Log.Fulfill(l)
+}
+
+// Log is the base logger used by kubebuilder.  It delegates
+// to another logr.Logger.  You *must* call SetLogger to
+// get any actual logging.
+var Log = log.NewDelegatingLogger(log.NullLogger{})
```
Comment (chuckha) [inline cmd/clusterctl/pkg/log/log.go]: Do we need the DelegatingLogger?

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/log/log.go]: I like the fact that the DelegatingLogger is explicitly providing a default null logger that works well with unit tests/as a default behavior from the library and it can also host user-provided loggers.

This can be achieved in other ways, but since we are already importing controller-runtime, I don't see backsides on taking benefit of this component as well

File: cmd/clusterctl/pkg/client/init.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: The former solution for handling this message was even worse,  with a boolean flag in output that was accepted during the review with the commitment to change this as soon as possible.

But I agree with your concern and this is why those logs are generated only if the caller explicitly requires it, and by default, they are not generated.

File: cmd/clusterctl/pkg/log/log.go (hunk 2 above)
Comment (chuckha) [reply cmd/clusterctl/pkg/log/log.go]: sgtm! thanks for your thoughts

**Comment by fabriziopandini** [comment]
@ncdc @vincepri 
Rebased; I have also split the changes in few commits hoping to simplify review

File: cmd/clusterctl/pkg/client/cluster/cert_manager.go [hunk 3]
```diff
@@ -65,6 +65,7 @@ func newCertMangerClient(proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *
 // Nb. In order to provide a simpler out-of-the box experience, the cert-manager manifest
 // is embedded in the clusterctl binary.
 func (cm *certMangerClient) EnsureWebHook() error {
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/cert_manager.go]:
```suggestion
func (cm *certManagerClient) EnsureWebHook() error {
```

File: cmd/clusterctl/pkg/client/cluster/inventory.go [hunk 4]
```diff
@@ -100,18 +100,24 @@ func newInventoryClient(proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *i
...
-			return errors.Wrap(err, "failed to check if the clusterctl inventory CRD exists")
-		}
+	if err = c.List(ctx, l); err == nil {
+		return nil
 	}
+	if !apimeta.IsNoMatchError(err) {
+		return errors.Wrap(err, "failed to check if the clusterctl inventory CRD exists")
+	}
+
+	// Otherwise install the cert-manager
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/inventory.go]:
```suggestion

```

**Comment by vincepri** [comment]
Going to unassign myself given that there are already 2 folks reviewing this

/unassign

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go [hunk 5]
```diff
@@ -231,14 +231,14 @@ func (o *objectGraph) Discovery(namespace string, types []metav1.TypeMeta) error
 			return errors.Wrapf(err, "failed to list %q resources", objList.GroupVersionKind())
 		}
 
-		o.log.V(2).Info(typeMeta.Kind, "Count", len(objList.Items))
+		log.V(5).Info(typeMeta.Kind, "Count", len(objList.Items))
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/objectgraph.go]: Does this log need a `msg` as the first argument?

File: cmd/clusterctl/pkg/client/cluster/proxy.go [hunk 6]
```diff
@@ -161,6 +161,10 @@ func (k *proxy) getConfig() (*rest.Config, error) {
 	}
 	restConfig.UserAgent = fmt.Sprintf("clusterctl/%s (%s)", version.Get().GitVersion, version.Get().Platform)
 
+	// Set QPS an Burst to a threshold that ensures the controller runtime client/client go does't generate throttling log messages
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/proxy.go]: nit: 
```suggestion
	// Set QPS and Burst to a threshold that ensures the controller runtime client/client-go doesn't generate throttling log messages
```

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 7]
```diff
@@ -79,24 +78,27 @@ func (o *objectMover) Move(namespace string, toCluster Client) error {
...
 func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+	log := logf.Log
+
 	clusters := graph.getClusters()
-	o.log.Info("Clusters to move", "Count", len(clusters))
+	log.Info("Moving Cluster API objects", "Clusters", len(clusters))
 
 	// Sets the pause field on the Cluster object in the source management cluster, so the controllers stop reconciling it.
-	if err := setClusterPause(o.fromProxy, clusters, true, o.log); err != nil {
+	log.V(1).Info("Stopping object reconciliation in the source cluster")
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/mover.go]:
```suggestion
	log.V(1).Info("Pausing the source cluster")
```

File: cmd/clusterctl/pkg/client/cluster/mover.go [hunk 8]
```diff
@@ -109,21 +111,24 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
...
+	log.Info("Deleting objects from the source cluster")
 	for groupIndex := len(moveSequence.groups) - 1; groupIndex >= 0; groupIndex-- {
 		if err := o.deleteGroup(moveSequence.getGroup(groupIndex)); err != nil {
 			return err
 		}
 	}
 
 	// Reset the pause field on the Cluster object in the target management cluster, so the controllers start reconciling it.
-	if err := setClusterPause(toProxy, clusters, false, o.log); err != nil {
+	log.V(1).Info("Starting object reconciliation in the target cluster")
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/mover.go]:
```suggestion
	log.V(1).Info("Resuming the target cluster")
```

File: cmd/clusterctl/pkg/client/init.go [hunk 9]
```diff
@@ -68,40 +71,53 @@ func (c *clusterctlClient) Init(options InitOptions) ([]Components, bool, error)
...
+		return nil, err
 	}
 
 	components, err := installer.Install()
 	if err != nil {
-		return nil, false, err
+		return nil, err
+	}
+
+	// IF this is the firstRun, then log the usage instructions.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]:
```suggestion
	// If this is the firstRun, then log the usage instructions.
```

File: cmd/clusterctl/pkg/client/init.go [hunk 10]
```diff
@@ -68,40 +71,53 @@ func (c *clusterctlClient) Init(options InitOptions) ([]Components, bool, error)
...
+	if firstRun && options.LogUsageInstructions {
+		log.Info("")
+		log.Info("Your cluster API management cluster has been initialized successfully!")
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]:
```suggestion
		log.Info("Your management cluster has been initialized successfully!")
```

File: cmd/clusterctl/pkg/client/delete.go [hunk 11]
```diff
@@ -70,12 +70,17 @@ func (c *clusterctlClient) Delete(options DeleteOptions) error {
 			}
 
 			// Check the provider/namespace tuple actually matches one of the installed providers.
+			found := false
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/delete.go]: The changes in this file seem like it should be in a separate PR.

File: cmd/clusterctl/pkg/log/log.go [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// SetLogger sets a concrete logging implementation for all deferred Loggers.
+func SetLogger(l logr.Logger) {
+	Log.Fulfill(l)
+}
+
+// Log is the base logger used by kubebuilder.  It delegates
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/log/log.go]: Just wanted to confirm. How is this `Log` used by kubebuilder?

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go (hunk 5 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/objectgraph.go]: No this is intentional. It reads:
Cluster Count=5
Machines Count=15
etc.etc.

File: cmd/clusterctl/pkg/client/delete.go (hunk 11 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/delete.go]: Yep, testing logs I found some small bugs and I fixed them.

**Comment by fabriziopandini** [comment]
@ncdc @wfernandes comment addressed! thanks for feedback!

File: cmd/clusterctl/pkg/client/cluster/objectgraph.go (hunk 5 above)
Comment (wfernandes) [reply cmd/clusterctl/pkg/client/cluster/objectgraph.go]: Oh...you are correct. I'm not sure what I was thinking.

File: cmd/clusterctl/pkg/client/client_test.go [hunk 13]
```diff
@@ -181,7 +181,7 @@ func (f fakeClusterClient) Proxy() cluster.Proxy {
 	return f.fakeProxy
 }
 
-func (f *fakeClusterClient) CertManger() cluster.CertMangerClient {
+func (f *fakeClusterClient) CertManger() cluster.CertManagerClient {
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/client_test.go]:
```suggestion
func (f *fakeClusterClient) CertManager() cluster.CertManagerClient {
```

File: cmd/clusterctl/pkg/client/client_test.go [hunk 14]
```diff
@@ -158,7 +158,7 @@ func newFakeCluster(kubeconfig string) *fakeClusterClient {
 type fakeCertMangerClient struct {
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/client_test.go]:
```suggestion
type fakeCertManagerClient struct {
```

File: cmd/clusterctl/pkg/client/client_test.go [hunk 15]
```diff
@@ -158,7 +158,7 @@ func newFakeCluster(kubeconfig string) *fakeClusterClient {
...
 }
 
-var _ cluster.CertMangerClient = &fakeCertMangerClient{}
+var _ cluster.CertManagerClient = &fakeCertMangerClient{}
 
 func (p *fakeCertMangerClient) EnsureWebHook() error {
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/client_test.go]:
```suggestion
func (p *fakeCertManagerClient) EnsureWebHook() error {
```

File: cmd/clusterctl/pkg/client/client_test.go [hunk 16]
```diff
@@ -158,7 +158,7 @@ func newFakeCluster(kubeconfig string) *fakeClusterClient {
 type fakeCertMangerClient struct {
 }
 
-var _ cluster.CertMangerClient = &fakeCertMangerClient{}
+var _ cluster.CertManagerClient = &fakeCertMangerClient{}
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/client_test.go]:
```suggestion
var _ cluster.CertManagerClient = &fakeCertManagerClient{}
```

File: cmd/clusterctl/pkg/client/cluster/cert_manager.go [hunk 17]
```diff
@@ -36,25 +36,25 @@ const (
...
-// Ensure certMangerClient implements the CertMangerClient interface.
-var _ CertMangerClient = &certMangerClient{}
+// Ensure certManagerClient implements the CertManagerClient interface.
+var _ CertManagerClient = &certManagerClient{}
 
-// newCertMangerClient returns a certMangerClient.
-func newCertMangerClient(proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *certMangerClient {
-	return &certMangerClient{
+// newCertMangerClient returns a certManagerClient.
+func newCertMangerClient(proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *certManagerClient {
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/cert_manager.go]:
```suggestion
func newCertManagerClient(proxy Proxy, pollImmediateWaiter PollImmediateWaiter) *certManagerClient {
```

File: cmd/clusterctl/pkg/client/cluster/client.go [hunk 18]
```diff
@@ -45,9 +44,9 @@ type Client interface {
 	// Proxy return the Proxy used for operating objects in the management cluster.
 	Proxy() Proxy
 
-	// CertManger returns a CertMangerClient that can be user for
+	// CertManger returns a CertManagerClient that can be user for
 	// operating the cert-manager components in the cluster.
-	CertManger() CertMangerClient
+	CertManger() CertManagerClient
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/client.go]:
```suggestion
	CertManager() CertManagerClient
```

**Comment by fabriziopandini** [comment]
@wfernandes comment addressed + squashed

**Comment by fabriziopandini** [comment]
/hold cancel
@wfernandes back to you

File: cmd/clusterctl/pkg/client/cluster/cert_manager.go [hunk 19]
```diff
@@ -110,10 +115,11 @@ func (cm *certMangerClient) EnsureWebHook() error {
 	}
 
 	// Waits for for the cert-manager WebHook to be available.
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/cert_manager.go]: not a blocker...but since we have the `log.Info` message I don't think we need this comment as well.
